### PR TITLE
RichText: remove Editable & allow React to diff editable element

### PR DESCRIPTION
**Pull request by ellatrix**
## Description

This PR eliminates our own diffing for attributes on the content editable element.

Benefits:

* No more expensive diffing of style and `aria-` attributes on every render.
* We normally have to manually add new attributes, e.g. `reversed` or `start` etc..
* No more `Editable` component.
* Gets rid of the `shouldComponentUpdate` hack.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate.

### CODE REVIEW

**Comment by aduth** [comment]
It seems really promising. What's the status of this? Just needs a review?

**Comment by ellatrix** [comment]
Yeah, this is ready for review :D

File: packages/rich-text/src/component/index.js [hunk 1]
```diff
@@ -988,6 +990,11 @@ class RichText extends Component {
 				onKeyUp={ this.onSelectionChange }
 				onMouseUp={ this.onSelectionChange }
 				onTouchEnd={ this.onSelectionChange }
+				role="textbox"
+				aria-multiline
+				aria-label={ placeholder }
```
Comment (aduth) [inline packages/rich-text/src/component/index.js]: We're changing behavior here: Previously, any `ariaProps` passed from props would always take priority, but now they're applied before we apply the defaults.

Should we spread `...ariaProps` after these instead?

File: packages/rich-text/src/component/index.js [hunk 2]
```diff
@@ -947,29 +956,22 @@ class RichText extends Component {
 
 	Editable( props ) {
```
Comment (aduth) [inline packages/rich-text/src/component/index.js]: While I understand a goal of this pull request was to eliminate an "Editable" component, that's effectively what I see this function as being, and perhaps we should still have it externalized from the parent?

One potential advantage could be in enabling you to use hooks.

**Review by aduth — CHANGES_REQUESTED** [review]
If I understand correctly, the changes here aren't actually leveraging React to do the rendering of the RichText record?

Even still, I think the benefits here are pretty great, and serve as a nice overall simplification.

I would be curious as well what the measurable performance difference is here, if any (seems there ought to be some improvement to key presses?).

Most of my review comments are minor nit-picks. Functionally, it works great. The only thing I'd consider really blocking is the priority with which we're applying ARIA props seems to have changed.

File: packages/rich-text/src/component/index.js [hunk 3]
```diff
@@ -947,29 +956,22 @@ class RichText extends Component {
...
 				{ ...props }
+				{ ...ariaProps }
 				ref={ forwardedRef }
-				tagName={ Tagname }
-				style={ style }
-				record={ this.record }
-				valueToEditableHTML={ this.valueToEditableHTML }
-				aria-label={ placeholder }
-				{ ...pickAriaProps( this.props ) }
+				style={ { ...style, whiteSpace } }
```
[the remark this answers is not in the document]
Comment (ellatrix) [reply packages/rich-text/src/component/index.js]: > Can we apply this style via CSS ? It avoids the whole need to merge the objects at all.

Oh you mean the white space rule? Yes, I considered using a stylesheet, but then this would be the only rule in the stylesheet, so I picked adding it to the style object instead of creating a whole new stylesheet for this package.

I'll adjust to check the style prop.

File: packages/rich-text/src/component/index.js (hunk 2 above)
Comment (ellatrix) [reply packages/rich-text/src/component/index.js]: The goal wasn't to eliminate `Editable`, but to have React manage the element. A side effect of that is that `Editable` is nothing more than just the element. What logic would move to the function in the form of hooks?

Comment (aduth) [reply packages/rich-text/src/component/index.js]: Well, I had a few initial thoughts around how we could `useMemo` for caching the combined style object (turned out maybe not viable, or at least the stylesheet idea could work better) or `useCallback` for the assigned value of `onKeyDown` (turned out if the component was made separate, this logic would have to occur in the parent `RichText` anyways).

tl;dr I wouldn't worry too much about it now, but maybe in the future there could be some additional requirements where hooks would prove to be useful.

**Comment by ellatrix** [comment]
> If I understand correctly, the changes here aren't actually leveraging React to do the rendering of the RichText record?

No, it's just letting React manage the editable element, not its contents. I'd like to explore that in the future, but I ran into some problems with it. React can't seem to update a DOM tree if the current tree is not in a shape that it's expecting. Anyway, that's a different problem.

> I would be curious as well what the measurable performance difference is here, if any (seems there ought to be some improvement to key presses?).

I think that, while there is a bit less work being done, the difference is negligible, especially compared to all the work that's being done outside of `RichText` after `onChange`.

File: packages/rich-text/src/component/index.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (ellatrix) [reply packages/rich-text/src/component/index.js]: I wonder if React has some internal optimisations for the `style` prop on an element. It's a bit strange to expect React users to optimise these style objects that are so commonly used.

File: packages/rich-text/src/component/index.js (hunk 2 above)
Comment (ellatrix) [reply packages/rich-text/src/component/index.js]: Yeah, I also agree that this component could use some more splitting in general. It's definitely on my mind.

File: packages/rich-text/src/component/index.js [hunk 4]
```diff
@@ -1024,6 +1041,8 @@ class RichText extends Component {
 				onKeyUp={ this.onSelectionChange }
 				onMouseUp={ this.onSelectionChange }
 				onTouchEnd={ this.onSelectionChange }
+				contentEditable
```
Comment (aduth) [inline packages/rich-text/src/component/index.js]: Technically we previously applied props such that someone could `contentEditable`. I don't think that's a use case we want to support though 😄

File: packages/rich-text/src/component/index.js [hunk 5]
```diff
@@ -58,6 +57,35 @@ const INSERTION_INPUT_TYPES_TO_IGNORE = new Set( [
...
+ * https://html.spec.whatwg.org/multipage/interaction.html#best-practices-for-in-page-editors
+ *
+ * @type {string}
+ */
+const whiteSpace = 'pre-wrap';
+
+/**
+ * Default style object for the editable element.
+ *
+ * @type {Object}
```
Comment (aduth) [inline packages/rich-text/src/component/index.js]: For future consideration: As a type, `Object` is better than nothing, but if we want to move toward better type enforcement (e.g. #17014), we probably want to be more specific about the sorts of values we're expecting to exist in the object.

```suggestion
 * @type {Object<string,string>}
```

In this case, it doesn't matter as much since it's a constant value ([aside: naming](https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/#constants)), but can be helpful for when possibly assigning into or consuming from an object to have the type awareness / enforcement.

**Review by aduth — APPROVED** [review]
Nice 👍

File: packages/rich-text/src/component/index.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (aduth) [reply packages/rich-text/src/component/index.js]: > I wonder if React has some internal optimisations for the `style` prop on an element. It's a bit strange to expect React users to optimise these style objects that are so commonly used.

Yeah, I'm not sure if they do, but as a general principle this is also why I tend to avoid objects as props in the first place. `ariaProps` is another good example of where this becomes needlessly difficult to manage. In retrospect, I'm not sure why we didn't just expect people to pass the specific prop they wanted to apply (`aria-foo`). I seem to recall some limitation that led us down that path, but certainly made things more difficult.

[the remark this answers is not in the document]
Comment (ellatrix) [reply packages/rich-text/src/component/index.js]: Right, I try to avoid them as much as possible too. Worth noting that `ariaProps` is picked from `props` though, there's no `ariaProps` prop. The picking seems to be there to avoid passing all `RichText` props to the element.

File: packages/rich-text/src/component/index.js (hunk 4 above)
Comment (ellatrix) [reply packages/rich-text/src/component/index.js]: Right. I don't think we should support that for now. We might in the future, maybe through a `disabled` or `readonly` prop. But probably not by directly setting `contentEditable.

File: packages/rich-text/src/component/index.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (aduth) [reply packages/rich-text/src/component/index.js]: > Right, I try to avoid them as much as possible too. Worth noting that `ariaProps` is picked from `props` though, there's no `ariaProps` prop. The picking seems to be there to avoid passing all `RichText` props to the element.

Oh! You're right, I definitely mis-recalled the behavior.

So maybe the opposite is true, that we have headaches because we chose to pick props than to accept an object 😄 Maybe in the future we could just be passing through all the props, so we don't need to be doing the explicit picking?

File: packages/rich-text/src/component/index.js (hunk 5 above)
Comment (ellatrix) [reply packages/rich-text/src/component/index.js]: I don't know that much about it. Why is `string` listed twice in `<string,string>`?

Comment (aduth) [reply packages/rich-text/src/component/index.js]: I find the following can be good for reference:

https://jsdoc.app/tags-type.html
https://www.typescriptlang.org/docs/handbook/type-checking-javascript-files.html#supported-jsdoc

There are two mentions of `string` to respectively represent the type of the key and the type of the value. (Which on reflection is a bit odd, since objects can only have string keys anyways)

File: packages/rich-text/src/component/index.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (ellatrix) [reply packages/rich-text/src/component/index.js]: I think we either need to pick the props we want to pass on, or delete the props we don't want to pass on. Both of these are explicit in a way I guess, but I'm fine with trying to delete the props we don't want to pass on. :) Either way there will be some sort of picking. We can't pass any unknown props to an element.

File: packages/rich-text/src/component/index.js (hunk 5 above)
Comment (ellatrix) [reply packages/rich-text/src/component/index.js]: Yeah, that is odd! :)

**Comment by ellatrix** [comment]
Thanks a lot for the review @aduth!

File: packages/rich-text/src/component/index.js (hunk 5 above)
Comment (aduth) [reply packages/rich-text/src/component/index.js]: Coming back to this, I discovered why it is useful to be able to define the type of the key. While true that an object key is always a string, we can do things like create enums of possible string options using a custom type.

See example at: https://github.com/WordPress/gutenberg/pull/18816#discussion_r353808652

Comment (ellatrix) [reply packages/rich-text/src/component/index.js]: Nice!
